dhall/src/Dhall/Parser/Expression.hs

```diff
@@ -181,16 +181,13 @@ parsers embedded = Parsers {..}
 
                     b <- expression
 
-                    case (shallowDenote a, shallowDenote b) of
-                        (ListLit _ xs, App f c) ->
-                            case shallowDenote f of
-                                List -> case xs of
-                                    [] -> return (ListLit (Just c) xs)
-                                    _  -> return (Annot a b)
-                                _ ->
-                                    return (Annot a b)
-                        (Merge c d _, e) ->
-                            return (Merge c d (Just e))
+                    case shallowDenote a of
+                        ListLit _ [] ->
+                            return (ListLit (Just b) [])
+                        ListLit _ _ ->
+                            return (Annot a b)
```

Comment:
This is the bit I was referring to from https://github.com/dhall-lang/dhall-haskell/pull/1112/files#r303272669.